designs/warmreplicas.md

```diff
@@ -0,0 +1,151 @@
+Add a new controller option `ShouldWarmupWithoutLeadership`. If set to true, then the main
+controller runnable will not start sources, and instead rely on the warmup runnable to start sources
+The option will be used as follows:
+```go
+type Controller struct {
+    // ...
+
+    // ShouldWarmupWithoutLeadership specifies whether the controller should start its sources
+    // when the manager is not the leader.
+    // Defaults to false, which means that the controller will wait for leader election to start
+    // before starting sources.
+    ShouldWarmupWithoutLeadership *bool
+
+    // ...
+}
+
+type runnableWrapper struct {
+    startFunc func (ctx context.Context) error
+}
+
+func(rw runnableWrapper) Start(ctx context.Context) error {
+    return rw.startFunc(ctx)
+}
+
+// NeedWarmup implements WarmupRunnable
+func (c *Controller[request]) NeedWarmup() bool {
+    if c.ShouldWarmupWithoutLeadership == nil {
+        return false
+    }
+    return c.ShouldWarmupWithoutLeadership
+}
+
+// GetWarmupRunnable implements WarmupRunnable
+func (c *Controller[request]) GetWarmupRunnable() Runnable {
+    return runnableWrapper{
+        startFunc: func (ctx context.Context) error {
+            if !c.ShouldWarmupWithoutLeadership {
+                return nil
+            }
+
+            // pseudocode
+            for _, watch := c.startWatches {
+                watch.Start()
+                // handle syncing sources
+            }
+        }
+    }
+}
+```
+
+3. Add a separate runnable category for warmup runnables to specify behavior when the
+manager is not the leader. [ref](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.2/pkg/manager/runnable_group.go#L55-L76)
+```go
+type runnables struct {
+    // ...
+
+	LeaderElection *runnableGroup
+    Warmup *runnableGroup
+
+    // ...
+}
+
+func (r *runnables) Add(fn Runnable) error {
+	switch runnable := fn.(type) {
+    // ...
+    case WarmupRunnable:
+        if runnable.NeedWarmup() {
+            r.Warmup.Add(runnable.GetWarmupRunnable(), nil)
+        }
+
+        // fallthrough to ensure that a runnable that implements both LeaderElection and
+        // NonLeaderElection interfaces is added to both groups
+        fallthrough
+	case LeaderElectionRunnable:
+		if !runnable.NeedLeaderElection() {
+			return r.Others.Add(fn, nil)
+		}
+		return r.LeaderElection.Add(fn, nil)
+    // ...
+	}
+}
+```
+
+4. Start the non-leader runnables during manager startup.
+```go
+func (cm *controllerManager) Start(ctx context.Context) (err error) {
+    // ...
+
+    // Start the warmup runnables
+	if err := cm.runnables.Warmup.Start(cm.internalCtx); err != nil {
+		return fmt.Errorf("failed to start other runnables: %w", err)
+	}
+
+    // ...
+}
+```
+
+## Concerns/Questions
+1. Controllers opted into this feature will break the workqueue.depth metric as the controller will
+   have a pre filled queue before it starts processing items.
```

Comment:
As discussed offline, one way to avoid this is to use a metrics wrapper that supresses them until the leader election is won. But not sure if its worth bothering

Reply:
Does this actually break the metric? Sounds like the metric will just show the reality

It might break alerts that assume the queue length should be pretty low, but that's an issue of the alerts.

Reply:
> It might break alerts that assume the queue length should be pretty low, but that's an issue of the alerts.

Not sure I quite agree with that. The alerts work today, if we change the behavior here, we break them. To my knowledge there also isn't a metric that indicates if a given replica is the leader, so I don't even see a good way to unbreak them

Reply:
Yeah, but the current definition of the metric is that it should show the length of the queue

Reply:
> To my knowledge there also isn't a metric that indicates if a given replica is the leader

That could be solved, I hope :)

Reply:
> Yeah, but the current definition of the metric is that it should show the length of the queue

Be that as it may, the reality is that there are a lot of ppl that use controller-runtime so [Hyrums Law](https://www.hyrumslaw.com/) applies - That we always have an empty workqueue when not leader is observable behavior and changing that will break ppl

Reply:
Then we maybe shouldn't store the items in the queue at this time because that's obsevable behavior as well (not only through the metric)

Reply:
> Then we maybe shouldn't store the items in the queue at this time because that's observable behavior as well (not only through the metric) and not just make it look like the queue is empty through the metric

How would that be obseveable except for through the metric if we don't start the controller?

Reply:
* logs via logState
* other workqueue metrics: adds_total, queue_duration_seconds
  * Although I guess we can also fake these. What would happen when the controller starts? I assume we would set the length metric immediately to it's correct value. Similar for adds_total and probably also queue_duration_seconds

I also think folks have programmatic access to the queue (at least by instantiating the queue in controller.Options.NewQueue)

So we don't know what kind of things folks are doing with the queue, e.g. accessing queue length or taking items out of the queue even if the leader election controllers are not running.

Reply:
> Then we maybe shouldn't store the items in the queue at this time because that's observable behavior as well (not only through the metric) and not just make it look like the queue is empty through the metric

That makes sense to me. I guess reading between the lines, the suggestion here is that non-leader runnables should The specific behavior I want to solve for is to have the queue ready to go asap after leader election, so would we implement something like "populate queue from cache on winning leader election"? I am not too familiar with the details here so would appreciate some concrete nudges in the right direction

Reply:
> That makes sense to me. I guess reading between the lines, the suggestion here is that non-leader runnables should The specific behavior I want to solve for is to have the queue ready to go asap after leader election, so would we implement something like "populate queue from cache on winning leader election"? 

I don't know how that would be possible, we would need to implement a dummy queue to buffer.

Lets punt on this problem until we make this the default behavior, we don't have to solve it right away. For as long as its opt-in and there is a metric indicating if a given replica is a leader, it won't break anyone by default and people can adjust their alerts accordingly.

Reply:
Sounds good!

Reply:
Thought more about this. Probably the initial suggestion is fine. We should just make sure that the metrics popping up once the controller becomes leader make sense